dev/TreeView/ViewModel.cpp

```diff
@@ -310,15 +318,85 @@ void ViewModel::NodeCollapsed(const winrt::event_token token)
 
 void ViewModel::SelectAll()
 {
+    BeginSelectionChanges();
+    auto trackSelection = gsl::finally([this]() { EndSelectionChanges(); });
+
     UpdateSelection(m_originNode.get(), TreeNodeSelectionState::Selected);
 }
 
-void ViewModel::ModifySelectByIndex(int index, TreeNodeSelectionState const& state)
+void ViewModel::SelectItem(winrt::IInspectable const& item)    
+{
+    BeginSelectionChanges();
+    auto trackSelection = gsl::finally([this]() { EndSelectionChanges(); });
+
+    auto selectedItems = GetSelectedItems();
+    if (selectedItems.Size() > 0)
+    {
+        selectedItems.Clear();
+    }
+    if (item)
+    {
+        selectedItems.Append(item);
+    }
+}
+
+void ViewModel::SelectNode(const winrt::TreeViewNode& node, bool isSelected)
+{
+    BeginSelectionChanges();
+    auto trackSelection = gsl::finally([this]() { EndSelectionChanges(); });
+
+    auto selectedNodes = GetSelectedNodes();
+    if (isSelected)
+    {
+        if (IsInSingleSelectionMode() && selectedNodes.Size() > 0)
+        {
+            selectedNodes.Clear();
+        }
+        selectedNodes.Append(node);
+    }
+    else
+    {
+        unsigned int index;
+        if (selectedNodes.IndexOf(node, index))
+        {
+            selectedNodes.RemoveAt(index);
+        }
+    }
+}
+
+void ViewModel::SelectByIndex(int index, TreeNodeSelectionState const& state)
 {
+    BeginSelectionChanges();
+    auto trackSelection = gsl::finally([this]() { EndSelectionChanges(); });
+
     auto targetNode = GetNodeAt(index);
     UpdateSelection(targetNode, state);
 }
 
+void ViewModel::BeginSelectionChanges()
+{
+    if (!IsInSingleSelectionMode())
+    {
+        m_selectionTrackingCounter++;
+        if (m_selectionTrackingCounter == 1) {
+            m_addedSelectedItems.get().Clear();
+            m_removedSelectedItems.get().Clear();
+        }
+    }
+}
+
+void ViewModel::EndSelectionChanges()
+{
+    if (!IsInSingleSelectionMode())
+    {
+        m_selectionTrackingCounter--;
+        if (m_selectionTrackingCounter == 0 && (m_addedSelectedItems.get().Size() > 0 || m_removedSelectedItems.get().Size() > 0)) {
```

Comment:
Is there a scenario where EndSelectionChanges could have been called and we don't have anything to report (i.e. both collections are empty)?

Reply:
Given the API exposed by the `ViewModel` type, yes, in theory.
You could make a call to `ViewModel::SelectItem` to select an already selected item.

I'm not sure this can happen given the code flows in the `TreeView`, I'm just extra careful.